### PR TITLE
CLI: Warn the user for duplicated versions after automigrate

### PR DESCRIPTION
**Pull request by yannbf**
Closes #

## What I did

When users migrate projects, they might end up having issues in Storybook due to duplicated versions of dependencies, which might clash in the major range. They could come as direct but also transitive dependencies. This PR adds a feature to the CLI's `automigrate` command to warn users when such scenario happens.

Here's an example **when there are duplicated versions for conflicting packages** in a pnpm project:

image

Here's an example **when there are also duplicated major versions** in a yarn2 project:
image

## How to test

1. Create a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Install conflicting versions e.g. (yarn add @storybook/addon-actions@6.0.0)
3. Run `code/lib/cli/bin/index.js automigrate` in that sandbox
4. See the warning

## Checklist

- [x] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [x] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

File: code/lib/cli/src/automigrate/index.ts [hunk 1]
```diff
@@ -381,10 +399,103 @@ function getMigrationSummary(
...
+    const duplicatedDependenciesMessage = Object.entries(
+      repoMetadata.duplicatedDependencies
+    ).reduce((acc, [dep, versions]) => {
+      if (allowList.includes(dep)) {
+        return acc;
+      }
+
+      const hasMultipleMajorVersions = hasMultipleVersions(versions);
+
+      if (hasMultipleMajorVersions && majorVersionClashMessage === '') {
```
Comment (tmeasday) [inline code/lib/cli/src/automigrate/index.ts:446]:
```suggestion
      if (hasMultipleMajorVersions) {
```

Not sure what the purpose of this check was?

File: code/lib/cli/src/automigrate/index.ts [hunk 2]
```diff
@@ -381,10 +399,103 @@ function getMigrationSummary(
...
+        majorVersionClashMessage = chalk.bold(
+          ' Given that some of these versions differ in the major range, there is a higher change Storybook will break.'
+        );
+      }
```
Comment (tmeasday) [inline code/lib/cli/src/automigrate/index.ts:450]: Should we add a message for when the disallow list is matched?

File: code/lib/cli/src/js-package-manager/parsePackageInfo.ts [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+};
+
+type PackageMetadata = { version: string; location?: string; reasons?: string[] };
+export type InstallationMetadata = {
+  dependencies: Record<string, PackageMetadata[]>;
+  duplicatedDependencies: Record<string, string[]>;
+  infoCommand: string;
+};
+
+export const mapDependenciesNpm = (input: NpmListOutput): InstallationMetadata => {
```
Comment (tmeasday) [inline code/lib/cli/src/js-package-manager/parsePackageInfo.ts:68]: Would it make sense to put this in `NPMProxy`? Is it reused between files?

**Review by tmeasday — COMMENTED** [review]
This is great! Is it possible to run the check without automigrating? As a like `sb doctor` type command? Or would you just suggest folks run `yarn sb migrate`?

File: code/lib/cli/src/automigrate/index.ts (hunk 1 above)
Comment (yannbf) [reply code/lib/cli/src/automigrate/index.ts:446]: The idea is to set that message only once

File: code/lib/cli/src/automigrate/index.ts (hunk 2 above)
Comment (yannbf) [reply code/lib/cli/src/automigrate/index.ts]: Done!

File: code/lib/cli/src/js-package-manager/parsePackageInfo.ts (hunk 3 above)
Comment (yannbf) [reply code/lib/cli/src/js-package-manager/parsePackageInfo.ts]: Done!

**Review by tmeasday — APPROVED** [review]
Great!

**Pull request merged**